packages/astro/src/core/config.ts

```diff
@@ -481,16 +477,36 @@ interface TryLoadConfigResult {
 async function tryLoadConfig(
 	configOptions: LoadConfigOptions,
 	flags: CLIFlags,
-	userConfigPath: string | undefined,
 	root: string
 ): Promise<TryLoadConfigResult | undefined> {
+	let finallyCleanup = async () => {};
 	try {
-		// Automatically load config file using Proload
-		// If `userConfigPath` is `undefined`, Proload will search for `astro.config.[cm]?[jt]s`
+		let configPath = (
+			await resolveConfigURL({ cwd: configOptions.cwd, flags: configOptions.flags })
+		)?.pathname;
+		if (!configPath) return undefined;
+		if (configOptions.isConfigReload) {
+			// Hack: Write config to temporary file at project root
+			// This invalidates and reloads file contents when using ESM imports or "resolve"
+			const tempConfigPath = path.join(
+				root,
+				`.temp.${Date.now()}.config${path.extname(configPath)}`
+			);
+			await fs.promises.writeFile(tempConfigPath, await fs.promises.readFile(configPath));
+			finallyCleanup = async () => {
+				try {
+					await fs.promises.unlink(tempConfigPath);
+				} catch {
+					/** file already removed */
+				}
+			};
+			configPath = tempConfigPath;
+		}
+
 		const config = await load('astro', {
-			mustExist: !!userConfigPath,
+			mustExist: true,
 			cwd: root,
-			filePath: userConfigPath,
+			filePath: configPath,
 		});
 
 		return config as TryLoadConfigResult;
```

Comment:
I might've approved too early. I tested locally, and whenever we update the config with invalid JS, the temp file isn't removed. I suppose we want to remove it in the catch block below this review line (github doesn't allow me to comment on it).

I nice cherry on top (though not necessary for this PR) is if the failed config load wouldn't crash the server, but only log the error and move on with the old astro config.

Reply:
Ah, probably because of the nested `try / catch` logic not running our cleanup correctly. Great catch!

Reply:
Alright, so fun fact: the error you're seeing is an **un-catch-able error.** Node file parse errors on imports bypass all try / catch logic, so if we want to delete that file at the right time, we have to get creative 🤔

Reply:
Update: discovered the issue in Proload. Making a patch over there to solve this! https://github.com/natemoo-re/proload/pull/29